### PR TITLE
feat(python): ADBC Support

### PR DESCRIPTION
**Pull request by owrior**
This replaces the read_sql functionality using connectorX to using ADBC. It seems faster, basic test on a table with 70K rows took 0.2s with ADBC which took 0.29s with connectorX. However it seems like ADBC doesn't currently support variable char length or date columns.

I'm happy to look at adding inserts etc but wanted to see if this is the sort of thing you had in mind for #6465

### CODE REVIEW

**Comment by owrior** [comment]
draft

**Comment by alexander-beedie** [comment]
Thanks for taking a look! Truthfully it's much too early for `ADBC` to _replace_ `connectorx` yet (as it only covers a small fraction of the supported databases and types, as you have discovered). However, hopefully it will advance quickly, and it looks like a great move for the Arrow ecosystem in general! Still, early days - I wouldn't look to do any more than have a configurable option for read-support at the moment (possibly set via `Config` as an experimental env var) until it gets closer to parity with features and database-support.

Now, we don't have _any_ write-support yet, so that may be a fruitful area to explore?

@ritchie46, any active plans in this area?
(It's not like you aren't already simultaneously working on 1001 other aspects of polars...😆)

**Comment by ritchie46** [comment]
I agree with @alexander-beedie on this one. I think it will always be a good hedge to support different sql engines under the hood.

I was hoping we could add an `engine: str` argument.

> Now, we don't have any write-support yet, so that may be a fruitful area to explore?
I have a lot of interest in arrow <-> database connectivity, so it's great to see progress here.

> @ritchie46, any active plans in this area?
(It's not like you aren't already simultaneously working on 1001 other aspects of polars...laughing)

No, none. This might also be something that patience and ADBC will give us. :)

**Comment by alexander-beedie** [comment]
> I was hoping we could add an `engine: str` argument.

@owrior: This sounds good - you could add the suggested `engine` param, default to "connectorx", but _also_ allow "adbc" for some initial/experimental support? Would be nice to be able to try it out easily, as it develops 👍 

> No, none. This might also be something that patience and ADBC will give us. :)

"Patience is a virtue..." 😁

**Comment by owrior** [comment]
Excellent sure thing will do! Thank you both.

File: py-polars/tests/unit/io/test_sql_db.py [hunk 1]
```diff
@@ -3,53 +3,181 @@
...
+
+        with pytest.raises(ValueError, match=err):
+            pl.read_sql(
+                connection_uri=f"{database}:///{test_db}",
+                sql=sql,
+                engine=engine,
+            )
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (alexander-beedie) [reply py-polars/tests/unit/io/test_sql_db.py:154]: Honestly I'd probably wait on writing; I don't think the maturity in this space is there yet. For example, from turbodbc's docs about Arrow support:

> _Apache Arrow support in turbodbc is still experimental and may not be as efficient as possible yet. Also, Apache Arrow support is not yet available for Windows and has some issues with Unicode fields. Stay tuned for upcoming improvements._

If there was a more fully-featured/mature package that we could rely on then I'd say go for it and put it behind an "engine" param, but I'm not currently not aware of one. Given that we may have to play a bit of a waiting game, unless you can find another one? 👍

[the remark this answers is not in the document]
Comment (ritchie46) [reply py-polars/tests/unit/io/test_sql_db.py:154]: >  should I add an alternative engine for writing? Like turbodbc for example?

Let's first get the ADBC in.

[the remark this answers is not in the document]
Comment (alexander-beedie) [reply py-polars/tests/unit/io/test_sql_db.py:154]: @owrior: BTW, fellow UCL grad here (Astrophysics, then CompSci) ;)

[the remark this answers is not in the document]
Comment (owrior) [reply py-polars/tests/unit/io/test_sql_db.py:154]: @alexander-beedie Oh cool! :) When did you graduate?

[the remark this answers is not in the document]
Comment (alexander-beedie) [reply py-polars/tests/unit/io/test_sql_db.py:154]: > @alexander-beedie Oh cool! :) When did you graduate?

_"A long time ago in a galaxy far, far away..."_  :)

[the remark this answers is not in the document]
Comment (owrior) [reply py-polars/tests/unit/io/test_sql_db.py:154]: XD

**Pull request closed without merging**